pytorch_lightning/metrics/metric.py

```diff
@@ -0,0 +1,122 @@
+
+
+def _sync_ddp(result: Union[torch.Tensor, numbers.Number],
+              group: Any = torch.distributed.group.WORLD,
+              reduce_op: torch.distributed.ReduceOp = torch.distributed.ReduceOp.SUM) -> torch.Tensor:
+    """
+    Function to reduce the tensors from several ddp processes to one master process
+
+    Args:
+        result: the value to sync and reduce (typically tensor or number)
+        group: the process group to gather results from. Defaults to all processes (world)
+        reduce_op: the reduction operation. Defaults to sum
+
+    Returns:
+        reduced value
+
+    """
+
+    # convert to tensor if necessary
+    if not isinstance(result, torch.Tensor):
+        result = torch.tensor(result)
+
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        # sync all processes before reduction
+        torch.distributed.barrier(group=group)
+        torch.distributed.all_reduce(result, op=reduce_op, group=group,
+                                     async_op=False)
+
+    return result
+
+
+def _sync_collections(result: Union[torch.Tensor, numbers.Number,
+                                    Mapping, Sequence],
+                      group: Any = torch.distributed.group.WORLD,
+                      reduce_op: torch.distributed.ReduceOp = torch.distributed.ReduceOp.SUM
+                      ) -> Union[torch.Tensor, numbers.Number,
+                                 Mapping, Sequence]:
+    """
+    Recursively applies sync_ddp to collections
+
+    Args:
+        result: Tensor or Number or Mapping or Sequence holding the values to be reduced
+        group: the process group to gather results from. Defaults to all processes (world)
+        reduce_op: the reduction operation. Defaults to sum
+
+    Returns:
+        the reduced collection
+
+    """
+    # function adapted from torch.utils.data._utils.collate
+    elem_type = type(result)
+
+    func = partial(_sync_collections, group=group, reduce_op=reduce_op)
+
+    # convert numpy to tensor if possible
+    if elem_type.__module__ == 'numpy' and elem_type.__name__ != 'str_' \
+            and elem_type.__name__ != 'string_':
+        # array not of string classes and object
+        if elem_type.__name__ != 'ndarray' \
+                or np_str_obj_array_pattern.search(result.dtype.str) is None:
+            result = torch.as_tensor(result)
+
+    if isinstance(result, (torch.Tensor, numbers.Number)):
+        return _sync_ddp(result, group=group, reduce_op=reduce_op)
+
+    elif isinstance(result, Mapping):
+        return elem_type({key: func(result[key]) for key in result})
+    elif isinstance(result, tuple) and hasattr(result, '_fields'):  # namedtuple
+        return elem_type(*(func(r) for r in result))
+    elif isinstance(result, Sequence) and not isinstance(result, str):
+        return elem_type([func(r) for r in result])
+    # not possible to reduce this type
+    else:
+        return result
```

Comment:
this is not safe since all `result` assignment is under some conditions...
rather instead each return do result assignment and always return result

Reply:
this should be safe, since we just return the input variable as is, if there is nothing to sync...